crypto/sleep.c

```diff
@@ -12,30 +12,37 @@
 
 /* system-specific variants defining OSSL_sleep() */
 #if defined(OPENSSL_SYS_UNIX) || defined(__DJGPP__)
-#include <unistd.h>
+# if defined(__TANDEM) && !defined(_REENTRANT)
 
+#  include <cextdecs.h(PROCESS_DELAY_)>
 void OSSL_sleep(uint64_t millis)
 {
-# ifdef OPENSSL_SYS_VXWORKS
-    struct timespec ts;
-
-    ts.tv_sec = (long int) (millis / 1000);
-    ts.tv_nsec = (long int) (millis % 1000) * 1000000ul;
-    nanosleep(&ts, NULL);
-# elif defined(__TANDEM) && !defined(_REENTRANT)
-#   include <cextdecs.h(PROCESS_DELAY_)>
-
     /* HPNS does not support usleep for non threaded apps */
     PROCESS_DELAY_(millis * 1000);
-# else
+}
+# elif defined(__DJGPP__)
+#  include <unistd.h>
+void OSSL_sleep(uint64_t millis)
+{
     unsigned int s = (unsigned int)(millis / 1000);
     unsigned int us = (unsigned int)((millis % 1000) * 1000);
 
     if (s > 0)
         sleep(s);
     usleep(us);
-# endif
 }
+# else
+/* nanosleep is defined by POSIX.1-2001 */
+#  include <time.h>
+void OSSL_sleep(uint64_t millis)
+{
+    struct timespec ts;
+
+    ts.tv_sec = (long int) (millis / 1000);
+    ts.tv_nsec = (long int) (millis % 1000) * 1000000ul;
+    nanosleep(&ts, NULL);
+}
+# endif
```

Comment:
Is this really supported on all our targets like AIX, HP/UX? I see this code will be used even on threaded NonStop builds. @rsbeckerca might want to look at this change.

Reply:
@t8m nanosleep is not broadly supported. This will break all of the NonStop distributions. I would suggest a separate knob in the Configuration defines to enable this, like `-DUSE_NANOSLEEP`, but retain the existing implementation. There are no plans as far as I can tell, to support `nanosleep()` on NonStop any time in the foreseeable future (min 5 years).

Reply:
It's a good question, and depends on each of them.

HP/UX uses pthreads (see hpux-common in Configurations/10-main.conf), so it's a pretty safe guess to say that they have enough POSIX to cover nanosleep() too.
One might make the same guess for AIX for the same reason.
For NonStop, I look forward to @rsbeckerca testing this!

Reply:
> It's a good question, and depends on each of them.
> 
> HP/UX uses pthreads (see hpux-common in Configurations/10-main.conf), so it's a pretty safe guess to say that they have enough POSIX to cover nanosleep() too. One might make the same guess for AIX for the same reason. For NonStop, I look forward to @rsbeckerca testing this!

Already tested, with bad results. Not supported on HPE NonStop nor is it in any plan I have seen. I can put an RFE in, but that will take years to percolate through the user base even if approved. I'm sorry.

Reply:
If we're implementing any `USE` macro, I'd rather have folks define `USE_USLEEP`, considering that usleep() is pronounced obsolete.  Let's not force `USE` macros for the commonly supported stuff, yeah?

For this case, though, we already detect `__TANDEM` for the no-REENTRANT stuff, and that's NonStop, isn't it?  Sooooo, how about this change, does that fix the NonStop issue?

``` diff
diff --git a/crypto/sleep.c b/crypto/sleep.c
index 4ce78c8462..d093e6dae3 100644
--- a/crypto/sleep.c
+++ b/crypto/sleep.c
@@ -20,7 +20,7 @@ void OSSL_sleep(uint64_t millis)
     /* HPNS does not support usleep for non threaded apps */
     PROCESS_DELAY_(millis * 1000);
 }
-# elif defined(__DJGPP__)
+# elif defined(__DJGPP__) || defined(__TANDEM)
 #  include <unistd.h>
 void OSSL_sleep(uint64_t millis)
 {
```

Reply:
The `PROCESS_DELAY_` is needed for unthreaded code. `usleep` is not supported outside of a thread mode on NonStop, so that has to be handled. We have `OPENSSL_SYS_TANDEM`. I have avoiding using `__TANDEM` on request from this team but I have no issue using the latter instead. As long as the `sleep/usleep` block of code is retained for this platform, and others that do not support `nanosleep()`, we should be fine.

Reply:
`__TANDEM` is already used in this source file, so I'd prefer to continue using that.  If someone wants to change that to `OPENSSL_SYS_TANDEM`, do feel free.